numpyro/contrib/nn/auto_reg_nn.py

```diff
@@ -86,107 +87,86 @@ class AutoregressiveNN(object):
         nonlinearity is applied to the final network output, so the output is an unbounded real number.
         defaults to ReLU.
     :type nonlinearity: callable.
+    :return: a tuple (init_fun, apply_fun)
 
     Reference:
 
     MADE: Masked Autoencoder for Distribution Estimation [arXiv:1502.03509]
     Mathieu Germain, Karol Gregor, Iain Murray, Hugo Larochelle
     """
-    def __init__(self, input_dim, hidden_dims, param_dims=[1, 1], permutation=None,
-                 skip_connections=False, nonlinearity=relu):
-        self.input_dim = input_dim
-        self.hidden_dims = hidden_dims
-        self.masked_layers = None
-        self.param_dims = param_dims
-        self.permutation = permutation
-        self.skip_connections = skip_connections
-        self.nonlinearity = nonlinearity
-        self.all_ones = (np.array(param_dims) == 1).all()
-
-        # Calculate the indices on the output corresponding to each parameter
-        ends = np.cumsum(np.array(self.param_dims), axis=0)
-        starts = np.concatenate((np.zeros(1), ends[:-1]))
-        self.param_slices = [slice(int(s), int(e)) for s, e in zip(starts, ends)]
-
-        self.count_params = len(self.param_dims)
-        self.output_multiplier = sum(self.param_dims)
-
-        # Hidden dimension must be not less than the input otherwise it isn't
-        # possible to connect to the outputs correctly
-        for h in self.hidden_dims:
-            if h < self.input_dim:
-                raise ValueError('Hidden dimension must not be less than input dimension.')
-
-    def init_fun(self, rng, input_shape):
+    output_multiplier = sum(param_dims)
+    all_ones = (np.array(param_dims) == 1).all()
+
+    # Calculate the indices on the output corresponding to each parameter
+    ends = np.cumsum(np.array(param_dims), axis=0)
+    starts = np.concatenate((np.zeros(1), ends[:-1]))
+    param_slices = [slice(int(s), int(e)) for s, e in zip(starts, ends)]
+
+    # Hidden dimension must be not less than the input otherwise it isn't
+    # possible to connect to the outputs correctly
+    for h in hidden_dims:
+        if h < input_dim:
+            raise ValueError('Hidden dimension must not be less than input dimension.')
+
+    net = None
+
+    def init_fun(rng, input_shape):
         """
         :param rng: rng used to initialize parameters
         :param input_shape: input shape
         """
-        if self.permutation is None:
+        # TODO: consider removing permutation so we can move those layer constructions outside
+        # init_fun. It seems that we can add a PermuteTransform layer to achieve the same effect.
+        nonlocal permutation, net
```

Comment:
is there no better solution than this `nonlocal` business?

Reply:
It is tricky I guess. We can do it if we remove permutation=None behavior (so users have to specify permutation unless using the default one). I think it is a good idea to remove permutation=None behavior. What do you think?

Reply:
The other solution would be to pass them back as return values in `init` and pass them back in in `apply`. Btw, there is little difference in behavior between using classes with fixed attributes (i.e. no change of state outside `__init__`), and using nonlocal in an initialization function to store and set variable values defined at the function scope level. The advantage of doing it this way is that it is easy to see where state setting happens and hard to accidentally mutate something outside of the initialization function. The advantage of using classes is that they are better supported, e.g. sphinx understands it well; likewise method signatures can be introspected easily by IDEs, pickling support for methods, etc.